### PR TITLE
Update Operations API

### LINKED ISSUE #804: Implement the (un)watching operations for not self-initiated operations

### LINKED ISSUE #806: Include watchlist features in operation dashboards

### LINKED ISSUE #812: Remove button for loading all/archived operations
- We don't have such long waiting times any more

### PR DESCRIPTION
**Pull request by bastihav**
# WIP
API tests need to be updated for new API refactor

# Description
Update Operations API to new workflow. Rejections are now frontend signing, endpoints have changed.

- Closes #806 
- Closes #804 
- Closes #812 

## Changes in this PR
- Refactor whole frontend signing behaivour of the API
- add functionality behind bookmark icons for watching and unwatching operations
- remove buttons for explicitly loading all/archived operations performance of the backend was improved significantly

## Type of change (remove all that don't apply)
- Feature
- Refactoring
- Breaking change (fix or feature that would cause existing functionality to not work as expected)

# How Has This Been Tested?

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration

- [X] Manually tested all use cases this could alter

**Test Configuration**:
- OS: Windows Mac Linux
- Browser: Firefox Chrome Safari Chromium

- Frontend: (remove all that don't apply)
  - [X] Development build
  
- Backend: (remove all that don't apply)
  - [X] deployed experimental build

# Checklist: (remove all that don't apply)

- [x] I added corresponding E2E tests (especially for bugfixes)
- [X] I added corresponding unit tests (if API code)
- [X] I have performed a self-review of my own code
- [X] My changes generate no new linting warnings or console warnings
- [x] I have updated the CHANGELOG.md to include any changes made in this PR (add WIP to the top, if there is none already)

### CODE REVIEW

File: src/api/CourseManagement.ts [hunk 1]
```diff
@@ -29,6 +34,9 @@ export default class CourseManagement extends Common {
         if (moduleIds != undefined) {
             requestParameter.params.moduleIds = moduleIds.reduce((a, b) => a + "," + b, "");
         }
+        if (examregNames != undefined) {
+            requestParameter.params.examregNames = examregNames.reduce((a, b) => a + "," + b, "");
```
Comment (tobias-wiese) [inline src/api/CourseManagement.ts:38]: Do we need the same fix here? (removing the last' ""')

File: src/api/helpers/OperationValidator.ts [hunk 2]
```diff
@@ -1,25 +1,24 @@
...
+        transactionInfo.contractName + ":" + transactionInfo.transactionName + ":" + transactionInfo.parameters.replace(/\s/g, "");
 
     const operationHash = await crypto.digest("SHA-256", new Uint8Array(toUTF8Array(toHash)));
 
     let operationId = arrayBufferToBase64(operationHash);
-
-    //url base64
     operationId = operationId.replace(/\+/g, "-");
     operationId = operationId.replace(/\//g, "_");
+    //operationId = operationId.replace(/=/g, "");
```
Comment (tobias-wiese) [inline src/api/helpers/OperationValidator.ts:19]:
```suggestion
```

File: src/api/helpers/UC4Identifier.ts [hunk 3]
```diff
@@ -7,7 +7,8 @@ export enum UC4Identifier {
 
     /* TRANSACTIONS */
     //common
-    TRANSACTION_APPROVAL = "approveTransaction",
+    TRANSACTION_APPROVAL = "approveOperation",
+    TRANSACTION_REJECTION = "rejectOperation",
```
Comment (tobias-wiese) [inline src/api/helpers/UC4Identifier.ts:11]: May rename the enum entry to `OPERATION_APPROVAL` etc

**Review by tobias-wiese — APPROVED** [review]
@bastihav you should have a look into the UI and test changes

File: src/api/helpers/OperationValidator.ts (hunk 2 above)
Comment (bastihav) [reply src/api/helpers/OperationValidator.ts:19]: We will need this with the next chaincode version, I am not sure if this branch will be merged before we test that version.

File: src/api/helpers/UC4Identifier.ts (hunk 3 above)
Comment (bastihav) [reply src/api/helpers/UC4Identifier.ts:11]: It is still a transaction to approve an operation. In this enum, all transactions are prefixed with transaction

File: src/api/CourseManagement.ts (hunk 1 above)
Comment (bastihav) [reply src/api/CourseManagement.ts:38]: No, this is a lagom bug and just "unexpected" use of the API, but still perfectly fine in the API definition

File: src/components/common/dashboard/Dashboard.vue [hunk 4]
```diff
@@ -183,7 +187,18 @@
                     })
                 );
 
+                //Get not-selfinitiated operations from watchlist
+                promises.push(
+                    operationManagement.getOperations(false, undefined, undefined, true).then((response) => {
+                        let result = handler.handleResponse(response);
```
Comment (bastihav) [inline src/components/common/dashboard/Dashboard.vue:193]:
```suggestion
                        const result = handler.handleResponse(response);
```

File: src/use/helpers/OperationPrinter.ts [hunk 5]
```diff
@@ -98,7 +98,8 @@ export function printOperationTitle(operation: Operation): string {
         case UC4Identifier.TRANSACTION_ADD_ENTRIES_MATRICULATION:
         case UC4Identifier.TRANSACTION_ADD_MATRICULATION:
         case UC4Identifier.TRANSACTION_UPDATE_MATRICULATION:
-            return "Update Matriculation";
+            var matrEntry = printMatriculationOperation(operation)[0].split(" ");
```
Comment (bastihav) [inline src/use/helpers/OperationPrinter.ts:101]:
```suggestion
            const matrEntry = printMatriculationOperation(operation)[0].split(" ");
```

File: src/views/admin/AllOperations.vue [hunk 6]
```diff
@@ -75,32 +77,34 @@
...
-                return result.id;
+                let ownId = "";
+                promises.push(
+                    certificateManagement.getOwnEnrollmentId().then((response) => {
+                        ownId = handler.handleResponse(response).id;
+                    })
+                );
+                await Promise.all(promises);
+                gotOps.value = true;
+                watchedOperations.value = tmpOps.filter((op) => op.initiator != ownId);
```
Comment (bastihav) [inline src/views/admin/AllOperations.vue:107]: correct me if I am wrong, but tmpOps is only used if you are not the initiator? So you can just use

`operationManagement.getOperations(false, undefined, undefined, true)`

the first parameter is selfInitiated

Comment (tobias-wiese) [reply src/views/admin/AllOperations.vue]: Removed in https://github.com/upb-uc4/ui-web/pull/826/commits/8b3cb54ee4dbed633cfbe1408263a925c3926695

**Pull request merged**